src/components/SignInButton.tsx

```diff
@@ -0,0 +1,212 @@
+import React, { useState } from 'react';
+import { isEmpty, isNil } from 'lodash/fp';
+import { Alert } from './Alert';
+import { ToS } from '../libs/ajax/ToS';
+import { DuosUser, User } from '../libs/ajax/User';
+import { Metrics } from '../libs/ajax/Metrics';
+import { Storage } from '../libs/storage';
+import { Navigation, setUserRoleStatuses } from '../libs/utils';
+import loadingImage from '../images/loading-indicator.svg';
+import { Auth } from '../libs/auth/auth';
+import eventList from '../libs/events';
+import { StackdriverReporter } from '../libs/stackdriverReporter';
+import CSS from 'csstype';
+import { useHistory } from 'react-router';
+import { OidcUser } from 'src/libs/auth/oidcBroker';
+
+interface SignInButtonProps {
+  customStyle: CSS.Properties | undefined;
+}
+
+interface ErrorInfo {
+  title?: string;
+  description?: string;
+}
+
+interface HttpError extends Error {
+  status?: number;
+}
+
+export const SignInButton = (props: SignInButtonProps) => {
+  const [errorInfo, setErrorInfo] = useState<ErrorInfo>({});
+  const [isLoading, setIsLoading] = useState<boolean>(false);
+  const [showError, setShowError] = useState<boolean>(false);
+  const { customStyle } = props;
+  const history = useHistory();
+
+  // Utility function called in the normal success case and in the undocumented 409 case
+  // Check for ToS Acceptance - redirect user if not set.
+  const checkToSAndRedirect = async (redirectPath: string | null) => {
+    // Check if the user has accepted ToS yet or not:
+    const user: DuosUser = await User.getMe();
+    if (!user.roles) {
+      await StackdriverReporter.report(
+        'roles not found for user: ' + user.email
+      );
+    }
+    setUserRoleStatuses(user, Storage);
+    const userStatus = await ToS.getStatus();
+    const { tosAccepted } = userStatus;
+    if (!isEmpty(userStatus) && !tosAccepted) {
```

Comment:
If `userStatus` was empty, wouldn't the destructuring in `const { tosAccepted } = userStatus;` throw an error?

When would we expect `userStatus` to be empty?

Reply:
> When would we expect `userStatus` to be empty?

I believe `userStatus` is one of the flavors of the Sam user info object. We collect that on every back end API call as part of the DUOS user to determine ToS acceptance.